google-cloud-debugger/lib/google/cloud/debugger/transmitter.rb

```diff
@@ -13,127 +13,184 @@
-          async_stop
+        #
+        # @return [Transmitter] returns self so calls can be chained.
+        def stop timeout = nil
+          if @thread_pool
+            @thread_pool.shutdown
+            @thread_pool.wait_for_termination timeout
+          end
+
+          self
         end
 
         ##
-        # Enqueue an evaluated breakpoints to be submitted by the transmitter.
+        # Whether the transmitter has been started.
         #
-        # @param [Google::Cloud::Debugger::Breakpoint] breakpoint The evaluated
-        #   breakpoint to be submitted
-        def submit breakpoint
-          synchronize do
-            @queue.push breakpoint
-            @queue_resource.broadcast
-            # Discard old entries if queue gets too large
-            @queue.pop while @queue.size > @max_queue_size
-          end
+        # @return [boolean] `true` when started, `false` otherwise.
+        #
+        def started?
+          @thread_pool.running? if @thread_pool
         end
 
         ##
-        # @private Callback fucntion for AsyncActor module to run the async
-        # job in a loop
-        def run_backgrounder
-          breakpoint = wait_next_item
-          return if breakpoint.nil?
-          begin
-            service.update_active_breakpoint agent.debuggee.id, breakpoint
-          rescue StandardError => e
-            warn ["#{e.class}: #{e.message}", e.backtrace].join("\n\t")
-            @last_exception = e
-          end
+        # Whether the transmitter has been stopped.
+        #
+        # @return [boolean] `true` when stopped, `false` otherwise.
+        #
+        def stopped?
+          !started?
         end
 
         ##
-        # @private Callback function when the async actor thread state changes
-        def on_async_state_change
-          synchronize do
-            @queue_resource.broadcast
-          end
+        # Register to be notified of errors when raised.
+        #
+        # If an unhandled error has occurred the transmitter will attempt to
+        # recover from the error and resume submitting breakpoints.
+        #
+        # Multiple error handlers can be added.
+        #
+        # @yield [callback] The block to be called when an error is raised.
+        # @yieldparam [Exception] error The error raised.
+        #
+        def on_error &block
+          @error_callbacks << block
         end
 
-        private
+        protected
 
-        ##
-        # @private The the next item from the transmitter queue. If there are
-        # no more item, it blocks the transmitter thread until an item is
-        # enqueued
-        def wait_next_item
-          synchronize do
-            @queue_resource.wait_while do
-              async_suspended? || (async_running? && @queue.empty?)
+        # Calls all error callbacks.
+        def error! error
+          error_callbacks = @error_callbacks
+          error_callbacks = default_error_callbacks if error_callbacks.empty?
```

Comment:
Why only add the defaults if the transmitter's `@error_callbacks` is empty? Would it be ok to add them to a non-empty `@error_callbacks` instead?

Reply:
If the transmitter does not specify any callbacks, the default ones will be used. The transmitter's callbacks are additive, there can be more than one. The config callback can be changed, and there can only be one. This allows the config's default callback to changed and be used while the previous config's default callback stops being used. This is the same pattern used in the other gems.